modules/imgcodecs/src/grfmt_gif.cpp

```diff
@@ -0,0 +1,372 @@
+            emptyImg.release();
+        } catch(...) {
+            return false;
+        }
+        return success;
+    }
+}
+
+void GifDecoder::readExtensions() {
+    uchar len;
+    while (!(m_strm.getByte() ^ 0x21)) {
+        uchar extensionType = (uchar)m_strm.getByte();
+
+        // read graphic control extension
+        // the scope of this extension is the next image or plain text extension
+        hasTransparentColor = false;
+        opMode = GifOpMode::GRFMT_GIF_Nothing;// default value
+        if (!(extensionType ^ 0xF9)) {
+            len = (uchar)m_strm.getByte();
+            CV_Assert(len == 4);
+            uchar flags = (uchar)m_strm.getByte();
+            m_strm.getWord(); // delay time, not used
+            opMode = (GifOpMode)((flags & 0x1C) >> 2);
+            hasTransparentColor = flags & 0x01;
+            transparentColor = (uchar)m_strm.getByte();
+        }
+
+        // skip other kinds of extensions
+        len = (uchar)m_strm.getByte();
+        while (len) {
+            m_strm.skip(len);
+            len = (uchar)m_strm.getByte();
+        }
+    }
+    // roll back to the block identifier
+    m_strm.setPos(m_strm.getPos() - 1);
+}
+
+void GifDecoder::code2pixel(Mat& img, int start, int k){
+    for (int i = start; i < height; i+=k) {
+        for (int j = 0; j < width; j++) {
+            int idx = i * width + j;
+            int colorIdx = currentImageCodeStream[idx];
+            if (hasTransparentColor && colorIdx == transparentColor) {
+                continue;
+            }
+            if (colorIdx < localColorTableSize) {
+                img.at<Vec3b>(top + i, left + j)[0] = localColorTable[colorIdx * 3 + 2]; //B
+                img.at<Vec3b>(top + i, left + j)[1] = localColorTable[colorIdx * 3 + 1]; //G
+                img.at<Vec3b>(top + i, left + j)[2] = localColorTable[colorIdx * 3];     //R
+            } else if (colorIdx < globalColorTableSize) {
+                img.at<Vec3b>(top + i, left + j)[0] = globalColorTable[colorIdx * 3 + 2]; //B
+                img.at<Vec3b>(top + i, left + j)[1] = globalColorTable[colorIdx * 3 + 1]; //G
+                img.at<Vec3b>(top + i, left + j)[2] = globalColorTable[colorIdx * 3];     //R
+            } else {
+                CV_Assert(false);
+            }
+        }
+    }
+}
+
+void GifDecoder::deleteLzwExtraTablePrefix(lzwNodeD* lzwExtraTable, int lzwTableSize) const{
+    for (int i = (1 << lzwMinCodeSize) + 2; i <= lzwTableSize; i++) {
+        if (lzwExtraTable[i].prefix) {
+            delete[] lzwExtraTable[i].prefix;
+            lzwExtraTable[i].prefix = nullptr;
+        }
+    }
+}
+
+bool GifDecoder::lzwDecode() {
+    // initialization
+    lzwMinCodeSize = m_strm.getByte();
+    int lzwCodeSize = lzwMinCodeSize + 1;
+    int clearCode = 1 << lzwMinCodeSize;
+    int exitCode = clearCode + 1;
+    CV_Assert(lzwCodeSize > 2 && lzwCodeSize <= 12);
+    auto* lzwExtraTable = new lzwNodeD[(1 << 12) + 1];
```

Comment:
please use std::vector, std::array or cv::AutoBuffer to prevent memory leaks in cases of parser failrue.

Reply:
done